pages/management/commands/upload_docx.py

```diff
@@ -0,0 +1,138 @@
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Command to extract text from either an uploaded document
+or a Google Docs link to set as a PageElement's text.
+"""
+import logging
+
+import os
+import re
+import zipfile
+from io import BytesIO
+from xml.etree.ElementTree import XML, ParseError
+
+import requests
+from django.core.management.base import BaseCommand, CommandError
+
+from ...models import PageElement
+
+LOGGER = logging.getLogger(__name__)
+
+
+WORD_NAMESPACE = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}'
+PARA = WORD_NAMESPACE + 'p'
+TEXT = WORD_NAMESPACE + 't'
+
+class Command(BaseCommand):
+    help = 'Uploads a .docx file from a local system or Google Drive link'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            'source',
+            type=str,
+            help='Source ("upload" for local file, "drive" for Google Drive link)')
+        parser.add_argument(
+            'identifier',
+            type=str,
+            help='File path or Google Drive link')
+        parser.add_argument(
+            'page_element_slug',
+            type=str,
+            help='Slug of the PageElement to update')
+
+    @staticmethod
+    def format_drive_url(url):
+        if 'docs.google.com/document' not in url:
+            raise CommandError(f'Invalid Google Docs URL: "{url}"')
+
+        match = re.search(r'/d/([0-9A-Za-z_-]{28,})/', url)
+        if not match:
+            raise CommandError(f'Could not extract document ID from Google Docs URL: "{url}"')
+
+        doc_id = match.group(1)
+        return f'https://docs.google.com/document/d/{doc_id}/export?format=docx'
+
+    @staticmethod
+    def get_docx_text(file_obj):
+        try:
+            with zipfile.ZipFile(file_obj) as document:
+                xml_content = document.read('word/document.xml')
+        except zipfile.BadZipFile:
+            raise CommandError(
+                'The file is not a valid zipfile (possibly not a docx).')
+
+        try:
+            tree = XML(xml_content)
+        except ParseError:
+            raise CommandError(
+                'There is an error parsing the document XML.')
+
+        paragraphs = []
+        for paragraph in tree.iter(PARA):
+            texts = [node.text for node in paragraph.iter(TEXT) if node.text]
+            if texts:
+                paragraphs.append(''.join(texts))
+
+        return '\n\n'.join(paragraphs)
+
+    def handle(self, *args, **options):
+        source = options['source']
+        identifier = options['identifier']
+        slug = options['page_element_slug']
+
+        if source not in ['upload', 'drive']:
```

Comment:
The `source` should most likely be derived from the `identifier`, i.e if the `identifier` starts with `http(s)` it is a Google Drive doc, otherwise it is a local file.